### PR TITLE
added wagner_fischer_distance

### PR DESCRIPTION
**Pull request by ruchir594**
Hello,

I added Wagner Fischer distance which is an algorithm to find edit distance using dynamic programming. Essentially gives the same result as Levenshtein but there is an advantage. 

With Wagner Fischer Distance, you can look at the matrix and see it is working. You can traverse the minimal change path to find out which exact alphabet in the string must be "edited". 

See the matrix, https://en.wikipedia.org/wiki/Wagner%E2%80%93Fischer_algorithm#Calculating_distance

### CODE REVIEW

**Comment by jamesturk** [comment]
Hi,

Thanks- if I understand correctly this will return the same value as Levenshtein?  If so I guess I'm not entirely clear on the use case?

File: .gitignore [hunk 1]
```diff
@@ -13,3 +13,4 @@ htmlcov/
 _build/
 .ipynb_checkpoints/
 .cache
+fortest.py
```
Comment (J535D165) [inline .gitignore]: Unintentionally added, isn't it?

**Comment by ruchir594** [comment]
As of today this will return the same value as Levenshtein. I am working on traversing the minimal change path to find out which exact alphabet in the string must be "edited" to what. So I want to return not only a number, but with alphabet locations on either string which must be edited.

Does this make sense?

**Comment by jamesturk** [comment]
Ah that does make sense.  I think it'd be best to wait to include this in
Jellyfish until it has a different use case from Levenshtein however.

**Comment by ruchir594** [comment]
Perfect. I will push more code soon once I finish the full use case. :)

**Comment by ruchir594** [comment]
Hello, 

I just added ability to "trace", by specifying it in the argument as trace = True. 

With this, now you know the edit-location in the edit-distance. This is done by traversing the matrix. I think it is a really helpful feature. 

You get to know if
1. the edits in the either string are sporadic, or aggregated into one location.
2. you can find patterns of edit location between words. 

for example, if you are comparing British english with American english words, you will find some word '~' is missing after 'o' which will be 'u'.
American English: Color, Favor 
British English: Colour, Favour 


jellyfish.levenshtein_distance(u'jellyfish', u'smellyfish')
jellyfish.wagner_fischer_distance(u'jellyfish', u'smellyfish')
jellyfish.wagner_fischer_distance(u'saturday', u'sunday', trace = True)

> python fortest.py 
levenshtein_distance 2
wagner_fischer_distance 2
wagner_fischer_distance (3, u's~~u~day')

**Comment by jamesturk** [comment]
Thanks- this is definitely interesting.  Out of personal curiosity if you
don't mind, what're you using this for?

Also, do you think it'd be possible to add tests to this PR in the style of
the other tests?

**Comment by ruchir594** [comment]
I am going to use this in our new Actions Architecture. http://tonatasha.com/ (Sorry, I haven't updated website in a long time. Drop me an email at ruchir594@gmail.com or ruchir@tonatasha.com and I am happy to talk more about it)
We are building this Architecture to build Super Smart Bots using NLU. We are using combination of pre trained ML models and contextual Latent Features. I am planning to include this in the latent feature layer somehow.  

And I will copy the levenshtein.csv and add a column for the traversed path. Cool?

File: jellyfish/_jellyfish.py [hunk 2]
```diff
@@ -218,6 +217,79 @@ def hamming_distance(s1, s2):
...
+    # filling up the rest of matrix by Flood Filling
+    for j in range(1, m):
+        for i in range(1, n):
+            if s1[i-1] == s2[j-1]:
+                d[i][j] = d[i-1][j-1]
+            else:
+                d[i][j] = min(d[i-1][j] + 1, \
+                              d[i][j-1] + 1, \
+                              d[i-1][j-1] + 1 )
+    if trace == False:
```
Comment (J535D165) [inline jellyfish/_jellyfish.py]: Maybe reverse this in favor of the argument ``None``. (And PEP8 doesn't like it) 

```
if trace:
    # code
else:
    return d[n-1][m-1]
```

File: jellyfish/_jellyfish.py [hunk 3]
```diff
@@ -218,6 +217,79 @@ def hamming_distance(s1, s2):
...
+                    traced = traced + s1[i-1]
+                    i = i - 1
+                else:
+                    traced = traced + '~'
+                    i = i - 1
+            elif point == d[i][j-1]:
+                traced = traced + '~'
+            j = j - 1
+        #drop last character
+        traced = traced[:-1]
```
Comment (J535D165) [inline jellyfish/_jellyfish.py]: These two lines are the same as:

`traced = traced[len(traced)-2::-1]`

**Comment by J535D165** [comment]
Maybe it is also nice to have the option to specify a trace character. Default '~'

```
>>> jellyfish.wagner_fischer_distance(u'saturday', u'sunday', trace = True, trace_char=".")

 (3, u's..u.day')
```

**Comment by ruchir594** [comment]
I will add the functionality to specify the "trace character". And will edit those changes! :D 

The checks failed because testdata/wagner_fischer.csv file is missing. I sent a pull request for that to the jellyfish-testdata git.

**Comment by ruchir594** [comment]
Hello,

Could someone accept the pull request at https://github.com/jamesturk/jellyfish-testdata to add wagner_fischer.csv to try the checks here? I hope my code in jellyfish/test.py works. Happy to receive further feedback.

**Comment by jamesturk** [comment]
Just accepted the PR.

Do you think there's a use here w/ trace=False?  Would it be better to just define this as returning the trace always so people don't get different numbers of items returned based on the param?

**Comment by ruchir594** [comment]
Failed because it cannot find the file 'wagner_fischer.csv': Does this mean you have to change the node of testdata in the Repo of Jellyfish? The current is at testdata @ 2c5b130.

**Comment by jamesturk** [comment]
If you go into the submodule and do a git pull, then a git add from outside of it, it'll add a commit that'll update the submodule.  If you can't get it working I can try to find time to do the update

**Comment by ruchir594** [comment]
I guess I will have to add code in cjellyfish in C? 

================================== FAILURES ===================================
____________________ test_wagner_fischer_distance[c---0-] _____________________
jf = <module 'jellyfish.cjellyfish' from 'C:\projects\jellyfish\jellyfish\cjellyfish.pyd'>
s1 = '', s2 = '', value = 0, traced = ''
    @pytest.mark.parametrize("s1,s2,value,traced", _load_data('wagner_fischer'), ids=str)
    def test_wagner_fischer_distance(jf, s1, s2, value, traced):
        value = int(value)
>       assert jf.wagner_fischer_distance(s1, s2)[0] == value
E       AttributeError: 'module' object has no attribute 'wagner_fischer_distance'
jellyfish\test.py:64: AttributeError
_________________ test_wagner_fischer_distance[c-abc--3-~~~] __________________
```

**Comment by ruchir594** [comment]
Update: for now, I have ignored the tests for CPython. :)

**Comment by J535D165** [comment]
If you are not going to make a C version, then you can change `__init__.py` into the following:

``` python 
try:
    from .cjellyfish import *   # noqa
    from ._jellyfish import wagner_fischer_distance
except ImportError:
    from ._jellyfish import *   # noqa
```

The Python and C version keep the same functionality. I think, there is no urgent need for a C version. Or are there large scale use cases? What is your opinion @jamesturk?

**Comment by ruchir594** [comment]
I am not planning to make a C version immediately. However, if there is a need for it, I am happy to do it.

**Pull request closed without merging**